### PR TITLE
Fixes #3262 Bad word filter

### LINKED ISSUE #3262: Bad word filter
Hello,
Simple problem, perhaps somebody can help me?

I have a word that I want to be censored on my forum. So I add the word in admin CP -&gt; configurations -&gt; Word Filters.

I add a word "whatever123" and it exchange into "space". So when somebody writes it it will change into an empty space.

So, when somebody types a message
"
Hi everybody,
My name is whatever123
I am a guy
Bye
"

It will be visible as:

"
Hi everybody,
My name is  
I am a guy
Bye
"

BUT!
When somebody types a message:
My name iswhatever123
OR
My name is whatever123I
OR
My name is whatever123Iam

The word "whatever123" will stay there. How can I get rid off it when its surrounded by other letters/words?
I thought I could add it to filter in the following way "#whatever123" or "%whatever123", but it doesnt work.

Any advice would be highly appreciated!
King regards!

Original thread: [Bad word filter](https://community.mybb.com/thread-218066.html)

### PR DESCRIPTION
**Pull request by effone**
Attempt to fix #3262 
Implemented effective usage of `*` and `+`

**Note: This is gonna break all the pre-defined bad word filters. Its is required to describe in release note to redefine bad word filters as per the new rule after applying this patch.**

### CODE REVIEW

**Comment by effone** [comment]
Note: Need to revise infinity loop check after this being merged. Here:
https://github.com/mybb/mybb/blob/feature/admin/modules/config/badwords.php#L50

I am not placing that modification now because doing that [this PR](https://github.com/mybb/mybb/pull/3355) will invariably introduce conflict.

**Comment by effone** [comment]
@euantorano PR completed. Thanks for sequential merge.

**Comment by euantorano** [comment]
Great, will test this one ASAP.

**Comment by Eldenroot** [comment]
Tested, seems to be working fine, good job!

File: inc/class_parser.php [hunk 1]
```diff
@@ -650,18 +650,12 @@ function parse_badwords($message, $options=array())
 					$badword['replacement'] = "*****";
 				}
 
-				if($badword['regex'])
```
Comment (Shade-) [inline inc/class_parser.php]: The whole PR works as expected, however, I'm not sure this is the intended behavior. IMHO, a filter with Regex option turned off should replace things without using regexes; so, if you specify `whatever123*` and Regex is off, it should replace the word literally.

Comment (Shade-) [reply inc/class_parser.php]: Briefly tested that, and it doesn’t work. Even with regex turned off, it also replaces words containing the filtered word, which is wrong.

Comment (effone) [reply inc/class_parser.php]: I honestly am not getting the issue. Can you please state steps with what intended and what is happening?

Comment (Shade-) [reply inc/class_parser.php]: Well, turns out I was testing in the wrong branch. It works as expected.

Comment (effone) [reply inc/class_parser.php]: Ha ha, thanks for the merge :D

**Comment by Eldenroot** [comment]
@effone - please check the last comment - abusing wordfilter by url tags, this will need a new issue and PR in future

**Comment by effone** [comment]
Noted.

File: inc/class_parser.php [hunk 2]
```diff
@@ -650,18 +650,12 @@ function parse_badwords($message, $options=array())
...
-					$message = preg_replace('#'.$badword['badword'].'#is', $badword['replacement'], $message);
+					$badword['badword'] = $this->generate_regex($badword['badword']);
 				}
-				else
-				{
-					// Take into account the position offset for our last replacement.
-					$badword['badword'] = str_replace('\*', '([a-zA-Z0-9_\*]{1})', preg_quote($badword['badword'], "#"));
 
-					// Ensure we run the replacement enough times but not recursively (i.e. not while(preg_match..))
-					$message = preg_replace("#(^|\W)".$badword['badword']."(?=\W|$)#i", '\1'.$badword['replacement'], $message);
```
[the remark this answers is not in the document]
Comment (effone) [reply inc/class_parser.php]: I don't get what you say here.

**Pull request merged**